Don't do BC check on ops with valid upgraders

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72313

Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

tugsbayasgalan committed:
Don't do BC check on ops with valid upgraders

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -109,6 +109,7 @@
     ("aten::nanquantile", datetime.date(2022, 9, 30)),
     ("aten::_convolution_double_backward", datetime.date(2022, 3, 31)),
     ("aten::_scatter_reduce", datetime.date(2022, 1, 31)),
+    ("aten::native_multi_head_self_attention", datetime.date(9999, 1, 1)),
```
I don't think we should add these indefinite windows to the allow list

tugsbayasgalan replied:
Sorry this is a bug. It shouldn't be here

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -137,6 +138,42 @@ def allow_listed(schema):
     ("dist_c10d", datetime.date(2099, 9, 17)),
 ]
 
+def has_valid_upgraders(schema, version_map):
```
Would you elaborate on the design thinking here? Is this just to avoid developers having to add an allow list entry when also writing an upgrader? Writing an allow list entry seems pretty easy.

tugsbayasgalan replied:
In version_map.cpp, we store the old schemas of the operators that are "upgraded". In this function, we just go over that list and make sure if the schema is one of those old signatures.

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

gmagogsfm commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,74 @@
+# Guidance for Operator Developer
+
+After you change either the operator signature in a BC breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking iff they are used in TorchScript or mobile. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `
```
Break this into multiple lines.

s/signature/schema/

s/your operator is BC breaking/your change to operator is BC-breaking

cccclai replied:
Resolve

gmagogsfm commented on docs/guidance_for_upgraders.md:
````diff
@@ -0,0 +1,74 @@
+
+    ```
+    ALL_MODULES = {
+        TestVersionedDivTensorExampleV7(): "aten::div.Tensor",
````
Do we have tests to verify that the key to this dictionary is a TestCase instance not a TestCase class? I think one of the common mistake is forgetting those parenthese.

cccclai replied:
Oh very good point! Let me have a pr to add test as well as error message

cccclai replied:
Address in https://github.com/pytorch/pytorch/pull/72615

gmagogsfm commented on docs/guidance_for_upgraders.md:
````diff
@@ -0,0 +1,74 @@
+    ```
+    3. Commit the change and submit a PR
+
+2. Write an upgrader in `torch/csrc/jit/operator_upgraders/upgraders_entry.cpp` file inside a map `kUpgradersEntryMap`. The softly enforced naming format is `<operator_name>_<operator_overload>_<start>_<end>`. For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need to be replaced by these upgraders. Note that the upgrader is essentially a python source ...
````
> For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need

I think this is technically correct, but there is no instruction on how the developer could find out what `0` and `7` should be.

gmagogsfm commented on docs/guidance_for_upgraders.md:
````diff
@@ -0,0 +1,74 @@
+        )SCRIPT"},
+```
+
+3. Bump the `kProducedFileFormatVersion` and `kMaxProducedFileFormatVersion` by 1 and provide the reasons under `caffe2/versions.h`.
````
Do we have tests to make sure that merge conflicts will not cause issue for this version bump?

cccclai replied:
Yeah it's a good point. We actually could run into merger conflicts in multiple files. We should add a section to explain how to merge.

gmagogsfm commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,74 @@
+
+3. Bump the `kProducedFileFormatVersion` and `kMaxProducedFileFormatVersion` by 1 and provide the reasons under `caffe2/versions.h`.
+
+4. In `torch/csrc/jit/operator_upgraders/version_map.cpp`, add changes like below. You will need to make sure that the entry is **SORTED** according to the version bump number. Otherwise, the sanity check will fail.
```
Sorted according to?

cccclai replied:
s/**SORTED** according to the version bump number/**SORTED** by the bumped to version number

gmagogsfm commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,74 @@
+This can be interpreted as: "We made a change to `div.Tensor` which caused a global operator version bump to 4
+. For outdated operators (aka 0 - 3) which have a signature `aten::div.Tensor(Tensor self, Tensor other) -> Tensor`, they need to use `div_Tensor_0_3` to be runnable in new runtime.
+
+5. After rebuild PyTorch, run the following command and it will auto generate a change to `torch/csrc/jit/mobile/upgrader_mobile.cpp`. This file contains the bytecode entries of the those upgraders which is used in mobile runtime.
```
s/rebuild/rebuilding/

gmagogsfm commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,74 @@
+# Guidance for Operator Developer
```
Overall, I think there are some grammar issues in this doc, could you try copy them into a google doc, which will detect these issues for you to fix.

cccclai replied:
Just did it and it seems fine. Will check again after the last commit

gmagogsfm commented on docs/guidance_for_upgraders.md:
````diff
@@ -0,0 +1,74 @@
+python pytorch/tools/codegen/operator_versions/gen_mobile_upgraders.py
+```
+
+6. Using the model generated from step 1, you will need to add tests in following places:
````
Instructions on how to add tests?

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

tugsbayasgalan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1761,6 +1761,7 @@ void initJitScriptBindings(PyObject* module) {
   m.def("_get_max_operator_version", &getMaxOperatorVersion);
   m.def("_get_operator_version_map", &get_operator_version_map);
   m.def("_get_upgrader_ranges", &getUpgradersRangeForOp);
+  m.def("_is_op_current", &isOpSymbolCurrent);
```
done

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -139,6 +137,42 @@ def allow_listed(schema):
+    # find if there are entries for this op
+    possible_overloads = []
+    possible_schemas = []
+    for key, upgrader_entries in version_map.items():
```
Shall we cache a "pre-processed" version_map? Its key should already be split for easier comparison, othewise we will be splitting them thousands of times for each test run

tugsbayasgalan replied:
sounds good

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -139,6 +137,42 @@ def allow_listed(schema):
+        return False
+
+    current_version = torch._C._get_max_operator_version()
+    for overload in possible_overloads:
```
Why do we need to check if if all possible_overloads is usable? Sounds like this isn't strictly necessary, is it because we can't tell which overload to match at this time?

tugsbayasgalan replied:
Yeah we can't tell which overload at this time, so we have to do it for all of them.

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

cccclai commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,74 @@
+
+# How to deprecate an upgrader?
+
+If an upgrader passes the expiration date and it’s a blocker for continuous development, it should be ok to remove it.
```
This part needs more details. I will remove this section for now and have a separate pr later to explain the process.

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -157,14 +194,33 @@ def load_schemas_to_dict():
+
+    output = defaultdict(dict)
+    for (key, entries) in version_map.items():
+        new_key = key.split(".")[0]
```
s/new_key/operator_name

tugsbayasgalan replied:
done

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -139,6 +137,45 @@ def allow_listed(schema):
+    # this existing schema can be up to date
+    if not found:
+        current_version = torch._C._get_max_operator_version()
+        for overload in possible_overloads:
```
Why do we need this for loop actually? If there is no upgrader (`if not found`) for it, then this `_is_op_symbol_current` check will always return `True` right?

tugsbayasgalan replied:
Yeah true. This is not needed.

tugsbayasgalan committed:
Update on "Don't do BC check on ops with valid upgraders"


Differential Revision: [D33997711](https://our.internmc.facebook.com/intern/diff/D33997711)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997711).

gmagogsfm commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -140,6 +138,33 @@ def allow_listed(schema):
+
+    # let's make sure this existing schema is part of possible
+    # schemas
+    for old_schema in possible_schemas:
```
Nit:
`return schema in possible_schemas`
